### DISCUSSION TITLE
Use Scala 3.3 LTS for SBT 2.x?

### DISCUSSION THREAD

**Discussion by mdedetrich** [sbt/sbt#7194] [Ideas]
Not sure if this was considered, but since we are discussing ideas for SBT 2 maybe we should also talk about Scala 3?

In my mind, there are some good arguments for using Scala 3 and more specifically Scala 3.3 LTS.

- There is a Scala 2 source flag which allows you to compile Scala 2 source code within Scala 3 and afaik this is not being removed in Scala 3.3 LTS. This would mean that as long as you use a modern Scala dialect you should be able to cross compile sbt plugins for 1.x (Scala 2) and 2.x (Scala 3) without too much trouble.
  - On this point, Scala 3.3 LTS is not adding that many source breaking changes. From my experience most of the changes were done in Scala 3.2 and they were pretty tame all things considered
 - The version that SBT supports kind of ends up creating a floor for the Scala version that the ecosystem ends up supporting, so using Scala 3 can help certain libraries not having to maintain really old Scala 2 versions
 - Scala 3.3 LTS includes some performance improvements (i.e. lazy val's, also Scala 3). While a minor point, anything that helps sbt perfromance wise is appreciated
 - Scala 3.3 being an LTS means that any future critical bug fixes/security issues will be backported to Scala 3.3 This allows SBT to continue using Scala 3.3 for a while, definitely ideal for such a critical piece of software
- Scala 3.3 LTS will also include the [stable-presentation-compiler-api](https://contributors.scala-lang.org/t/stable-presentation-compiler-api/6139). What this means is that tooling that will use it (i.e. metals and all the editors that use metals) will be able to support Scala 3.3 LTS and hence sbt builds for a very long time
- While the Scala 2 macros will have to be rewritten for Scala 3, the advantage Scala 3 brings it that it has much better compiler errors/warnings/diagnostics for macros, again improvement the experience for end users.

**Comment by mzuehlke** [comment]
The `sbt-2.x` branch developed by @eed3si9n  is using Scala 3 😄 
https://github.com/sbt/sbt/pull/6746

**Reply by mdedetrich** [reply]
Oh perfect then! I guess we should discuss using Scala 3.3 LTS specifically then (let me rename discussion).

**Comment by SethTisue** [comment]
Definitely agree that for build definitions we should not target anything under 3.3, once we have a 3.3.x release that appears stable.

Dropping 3.0/3.1/3.2 as soon as reasonably possible is advice I would give to pretty much any Scala OSS project, really. Even those beginning with P 😉

**Comment by eed3si9n** [comment]
Should we stick with Scala 3.3 long after 3.4 and 3.5 comes out too? Wouldn't that effectively hold back the plugin ecosystem and its transitive dependencies to an "old" Scala version?

**Reply by SethTisue** [reply]
It's much less clear what's going to happen there :-/

**Reply by mdedetrich** [reply]
> Should we stick with Scala 3.3 long after 3.4 and 3.5 comes out too? Wouldn't that effectively hold back the plugin ecosystem and its transitive dependencies to an "old" Scala version?

Unless I misunderstood, wouldn't SBT sticking to 3.3 LTS effectively create a floor? Scala programs compiled with versions equal to or greater than 3.3 can use 3.3 so SBT sticking with 3.3 for a while shouldn't be an issue?

Taking this from https://www.scala-lang.org/blog/2022/08/17/long-term-compatibility-plans.html, i.e.

>Scala 3 guarantees backward compatibility between all releases and forward compatibility between patch releases in the same minor line. That means that Scala 3.b.y can consume the output of Scala 3.a.x only if b is greater or equal to a. For example, the output of the Scala 3.0.2 compiler can be used by Scala 3.0.1 or 3.1.0. On the other hand, code compiled with Scala 3.1.0 cannot be a dependency of any project compiled with 3.0.2.

If this is correct then the side effect of using 3.3 is that you would force the rest of the ecosystem to support at least 3.3

**Reply by eed3si9n** [reply]
I think it means Scala 3.3 is not forward compat with 3.4 ecosystem. So if Cats or whatever goes to 3.4 we won't be able to use the latest in the metabuild? (I think)

**Reply by mdedetrich** [reply]
> I think it means Scala 3.3 is not forward compat with 3.4 ecosystem. So if Cats or whatever goes to 3.4 we won't be able to use the latest in the metabuild? (I think)

I have a different understanding then. I think this is the case for source compatibility, not binary.

Also in the previously link it does say its highly recommended that library authors/commercial authors should stick to LTS which wouldn't make sense if it broke in the way you describe.

This should be clarified.

**Reply by sjrd** [reply]
Because sbt loads plugins via run-time reflection in a child classloaders, all the plugins effectively inherit the scala-library of the running sbt instance. If sbt uses scala-library 3.3.x, the plugin will get scala-library 3.3.x on its classpath, no matter the version with which it was compiled.

*Forward* binary compatibility is not guaranteed across 3.x versions. If the sbt plugin uses a method added in scala-library 3.4.x, it will blow up at run-time.

So yes, this *will* hold back sbt plugins to the LTS Scala version.

Is that an issue? I mean, sbt plugins will probably want to still support older sbt's for a while, so they will target that "old" LTS Scala version anyway?

Conversely, upgrading to newer Scala in newer sbt is probably not an issue either. It does mean that a Scala shop with a large sbt codebase (and by that I mean build + custom sbt plugins; not the code being built) may have trouble upgrading to newer versions of sbt, hence newer fixes. But do we expect situations like that (large sbt codebase) to really happen?

**Reply by mdedetrich** [reply]
> Forward binary compatibility is not guaranteed across 3.x versions. If the sbt plugin uses a method added in scala-library 3.4.x, it will blow up at run-time.

Are you talking about sbt's case specifically due to its usage of classloaders/reflection or in general?

Also I was under the impression that for Scala 3 they don't want to add new methods (at least to the stdlib if that's what you were implying) in order to avoid this issue?

**Reply by sjrd** [reply]
> > Forward binary compatibility is not guaranteed across 3.x versions. If the sbt plugin uses a method added in scala-library 3.4.x, it will blow up at run-time.
> 
> Are you talking about sbt's case specifically due to its usage of classloaders/reflection or in general?

About sbt specifically, because of classloaders, which is a typical case of "plugin" architecture.

> Also I was under the impression that for Scala 3 they don't want to add new methods (at least to the stdlib if that's what you were implying) in order to avoid this issue?

No that is not true. Every new minor version of Scala 3 adds new things in the stdlib. (there are no changes in *patch* versions, though, to preserve forward binary compatibility within a patch series).

**Reply by mdedetrich** [reply]
> No that is not true. Every new minor version of Scala 3 adds new things in the stdlib. (there are no changes in patch versions, though, to preserve forward binary compatibility within a patch series).

Thanks for clarifying, must have misread this somewhere.

**Reply by adpi2** [reply]
New minor versions of Scala 3 add new things in the scala3-library which is a complement of the scala-library_2.13 for Scala 3.

**Comment by eed3si9n** [comment]
2.0.0-M2 was compiled against the 3.3.x LTS series, but I'm now sending a PR to bump it to "Scala Next" series 3.5.x - https://github.com/sbt/sbt/pull/7834.

Why the change of heart? Here are some reasons:
1. There a few semantic / syntactic changes that made sbt/sbt not even compile on Scala 3.5.x as-is. Part of the reasons on working on sbt 2.x is to bring the code base up to date, so it would defeat the purpose for sbt 2.x project (at least during the development) to settle for LTS prematurely.
2. sbt, even during sbt 0.13 era has operated with **backward-compatibility only** with regards to the plugin compatibility. Meaning, if you publish a plugin against sbt 1.10.3, it probably won't work in sbt 1.5.8 etc. This actually matches the backward-compat only nature of the Scala 3.x minor series. So let's say we ship sbt 2.0.0 with Scala 3.5.x, and sbt 2.1.0 with Scala 3.6.0. That means that a plugin published using sbt 2.10 will not work on sbt 2.0.0. But this is fine, because it wasn't going to regardless.
3. Similar to sbt 1.x should not hold back the Scala ecosystem to 2.12 world, sbt 2.x should not hold back permanently holdback the ecosystem (at least any libraries used by the plugins to 3.3.x LTS). For example, if someone wants to use Cats or Toolkit or whatever, and if sbt 2.x stayed on 3.3.x forever, they won't be able to use the Cats or Toolkit published against 3.4+.

**Update**: Added "3.", repeating the point I made in one thread above.

**Reply by SethTisue** [reply]
This might be confusing to users. instead of "3.3 LTS is for libraries", the new story would be "3.3 LTS is for libraries, except for sbt plugins, who went their own way on this"

I'm not necessarily opposed, just exploring options.

As for making sure that the code compiles on 3.5, a CI job could ensure that, while still keeping 3.3 as the actual version used for the rest of CI and for publishing.

> There a few semantic / syntactic changes that made sbt/sbt not even compile on Scala 3.5.x as-is.

(But we can keep those changes and still stay on 3.3.x, can't we?)

**Reply by eed3si9n** [reply]
@SethTisue 

> This might be confusing to users. instead of "3.3 LTS is for libraries", the new story would be "3.3 LTS is for libraries, except for sbt plugins, who went their own way on this"

There's a bit of mental gymnastics that's involved because sbt main is an app / framework, but we can think of the _plugins_ to be libraries. For example, sbt-assembly can be built against 3.3 LTS, and that will be guaranteed to work on all sbt 2.x series as long it uses any Scala 3.x above 3.3. But if sbt 2.x decided to stay up on 3.3 LTS, then Play etc will never be able to use anything above 3.3 LTS. So sbt main should try to stay together with the evolution of Scala 3 Next, but plugin could stay as low Scala version to maximize the range of sbt 2.x versions it can support (assuming it would also live long).

Added "3." point in the thread.

**Reply by adpi2** [reply]
If we bump the Scala version regularly it is going to fail the Scala Steward updates in many builds, because of the stricter syntax. And builds will tend to stay longer on old sbt versions.

Also, according to the [Long Term Compatibility Plans blog post](https://scala-lang.org/blog/2022/08/17/long-term-compatibility-plans.html#our-solution-scala-lts--scala-next), `3.3` won't stay LTS forever.  In the future, there will be another `3.X` LTS, and that could be the signal for upgrading sbt and all its ecosystem.

It's tempting to use the latest Scala version, but maybe we should blindly follow the recommendation. Should I bring this discussion to the attention of the Scala core team?

> But if sbt 2.x decided to stay up on 3.3 LTS, then Play etc will never be able to use anything above 3.3 LTS

We can change our mind when this will occur, if it will.

**Reply by mdedetrich** [reply]
At the risk of missing something, its also my understanding that sbt and all of its plugins should stick to Scala 3.3.x LTS and not follow scala-next.

Even bumping to the next Scala 3 LTS should be done with extreme caution, as you are basically going to brake all of the sbt plugins

**Reply by adpi2** [reply]
> For example, if someone wants to use Cats or Toolkit or whatever, and if sbt 2.x stayed on 3.3.x forever, they won't be able to use the Cats or Toolkit published against 3.4+.

Alternatively we can continue publishing sbt on 3.3 LTS, but force sbt to start with 3 NEXT.

**Reply by mdedetrich** [reply]
> For example, if someone wants to use Cats or Toolkit or whatever, and if sbt 2.x stayed on 3.3.x forever, they won't be able to use the Cats or Toolkit published against 3.4+.

On this point, a library like cats really should **not** be targeting Scala NEXT unless there is a very strong reason otherwise (i.e. they **need** a feature from Scala NEXT, and even then they should release a separate artifact for this). Libraries should be sticking to Scala 3.3.x LTS, thats what Pekko does and its for very good reason because it creates a floor of the minimum supported Scala 3 version and if you have your Scala 3 version too high then you create the situation where users of your library can't actually use it because they are on a lower Scala version.

For appications its a bit different but Cats is not an application.

**Reply by SethTisue** [reply]
> There's a bit of mental gymnastics that's involved

That's exactly what concerns me — just as much as the concrete practical consequences concern me. Mental gymnastics are an ecosystem-wide cost.

**Reply by SethTisue** [reply]
> On this point, a library like cats really should not be targeting Scala NEXT 

Agree. And I would add that in my experience, the vast majority of libraries are _in fact_ targeting LTS.

Yes, I've seen a few exceptions where someone really needed some fix from Scala Next, and I also have seen a few occasions where a library author unknowingly upgraded, but then downgraded again when a user complained.

But these exceptions are rather rare.

**Reply by eed3si9n** [reply]
@adpi2 wrote:

> It's tempting to use the latest Scala version, but maybe we should blindly follow the recommendation. Should I bring this discussion to the attention of the Scala core team?

Yes. I am kind of curious what Scala core team would recommend sbt and its hundred plugins /cc  @Kordyjan

https://scala-lang.org/development/ states:

> - The **Scala Next** line is the default to be used by most users, containing the latest features, bug fixes and improvements.
> - The **Scala LTS** line is advised to be used for publishing **libraries**. (Some especially conservative users might also choose it over Scala Next.)

In my opinion, sbt is not a library, it's an application. From the point of view of a plugin, a plugin would depend on sbt main, but from the perspective of the build users, you run sbt as an application, while loading sbt plugins as libraries to your source-less "metabuild", which is blended into the application. So if we want to maximize the flexibility for the plugin authors, I feel like picking the latest and greatest Scala 3.x version basically allows them to use both new language features and allows transitive libraries to move forward.

> If we bump the Scala version regularly it is going to fail the Scala Steward updates in many builds, because of the stricter syntax.

This is a good point actually. If Scala Next continues to break source compatibility, that might be a good argument not adopt while a release is in Next phase, and only change every two years when LTS comes out.

@mdedetrich 

> Even bumping to the next Scala 3 LTS should be done with extreme caution, as you are basically going to [break] all of the sbt plugins

As long as sbt 2.x maintains backward compatibility, previous releases of plugins would continue to work. So are you talking about source-compat breakages? If the trend of sbt 1.x continues, the plugins would settle on some early version (like sbt 1.2.8, or 1.5.8), so even that would be somewhat gradual, in a sense that some plugin that really need to bump to the latest sbt 2.x would need to make some source changes to adopt to Scala 3.x changes.

When the next LTS comes out (maybe in 2025 after two years from 2023?), hopefully the situation would become clearer. Would the libraries adopt the 2025 LTS etc. In general, my overall point I want to call out is that being stuck on Scala 3.3.0 LTS forever would be an unhealthy move for the project. Just as same as being on Scala 2.10 forever was not good.

@SethTisue 

> That's exactly what concerns me — just as much as the concrete practical consequences concern me. Mental gymnastics are an ecosystem-wide cost.

Don't blame the gymnast, blame the Olympics. The situation is a byproduct of backward-compat scheme. Also an LTS would go away in three years. With sbt 1.x, we usually bump to the latest Scala 2.12.x immediately for JDK support etc. I want to maintain that level of freshness for sbt 2.x.

**Reply by adpi2** [reply]
@eed3si9n 

> Also an LTS would go away in three years. With sbt 1.x, we usually bump to the latest Scala 2.12.x immediately for JDK support etc. I want to maintain that level of freshness for sbt 2.x.

An LTS patch is released every 3-6 months. Not sure if JDK support is in the scope, but probably since it is a "usability feature".

> So if we want to maximize the flexibility for the plugin authors, I feel like picking the latest and greatest Scala 3.x version basically allows them to use both new language features and allows transitive libraries to move forward.

It is even better if we stick to 3.3 LTS and allow sbt to launch with NEXT. We don't force anyone to bump to NEXT, not even the build files, but we allow plugins, libraries and builds to use NEXT if they want to. It seems LTS is more flexible for everyone, except us the sbt maintainers.

**Reply by mdedetrich** [reply]
> An LTS patch is released every 3-6 months. Not sure if JDK support is in the scope, but probably since it is a "usability feature".

Yup, and iirc its also backwards **and** forwards compatible with every Scala 3.3.x version since it follows semver

>  It seems LTS is more flexible for everyone, except us the sbt maintainers.

Completely agree with this, launching sbt with NEXT is acceptable although I am still wary in that it may encourage and/or give the wrong idea to sbt plugins which should stick to Scala 3.3.x LTS.

I also don't completely understand the core issue here, SBT 1.x Stuck with Scala 2.12.x for over a decade? Its not like SBT was targetting, 2.11, 2.12 and 2.13 which is what similarly would happen if you allowed sbt plugins to target NEXT. Its just going to open the floodgates, because you also have sbt plugins depending on other sbt plugins and you will get so many incompatibilities here as they will all have to cross compile to multiple Scala NEXT versions.

The path to tht least amount of chaos and ironicially flexibility is to just freeze (and also enforce if possible) Scala 3.3.x LTS for all sbt 2.x plugins, the launcher/application itself can use NEXT if it want's to but I don't know what this is really hoping to achieve (I guess some new syntax/macros etc etc might make things nicer for sbt but then again it can only be used for sbt itself and these features cannot be exposed to any of the plugins in any api/interface).

**Reply by tgodzik** [reply]
I don't think the core team will be able to give a definite answer here. The intuitive answer would be to keep sbt on LTS, but it's not something we would want to force on any contributors. We basically have two options here:
- stay on LTS and jump every two years (we expect the next LTS to be late next year) <- super safe approach
- keep bumping to Next to keep the sbt up to date with language features <- more fun approach

Personally, I think using Next would be ok. I don't believe sbt plugins use code that could easily be broken by the changes in Scala 3 Next. From my experience tooling doesn't use a lot of fancy abstractions (most of the time). And besides we do run the community build on around 1750 and even with the newest 3.6.x the breakages are rather small, related mostly to one change and easily fixable. We could also add sbt and plugins that migrate to the community build and let people know if there are indeed any breakages. Most plugins are open source in contrast to typical applications, so we would discover most problems before even a new Next version is fully released.

**Reply by sjrd** [reply]
The main danger of using Next for sbt is the following scenario:
* sbt upgrades to new minor Scala
* A user's build does not compile with the newest Scala (deprecated features became errors, for example)
* That user is an LTS user of Scala, because they cannot easily upgrade their Scala version
* The same reason applies to their build
* Therefore they cannot upgrade sbt
* (time passes)
* A "bad event" happens: new JDK for which we need support; or security fix in sbt/Scala; etc.
* New sbt version gets released with a fix for the "bad event"; new Scala LTS as well
* User still cannot upgrade sbt, although they could upgrade their Scala LTS
* User is now stuck on old JDK, or vulnerable to a security issue, etc.

This is the main reason we have an LTS of Scala: so that fixes that should really reach everyone *can* reach everyone. If we use Next in sbt, we are breaking that property.

If sbt itself had an LTS that would use Scala LTS, that would be mitigated, but that is not the case, and I don't think there are enough resources on sbt to be able to do that.

If sbt stays on LTS, none of these issues can happen. sbt plugins need to stay on LTS for that property to hold, but that is no different than the fact that libraries in general should stay on LTS.

**Reply by eed3si9n** [reply]
For the LTS option there are two proposed approaches:
- **LTS Option A**: Stay on Scala LTS 2023 (3.3.x) forever
- **LTS Option B**: Move with LTS, so LTS 2025 comes out, bump to LTS 2025

LTS has a planned lifecycle (planned obsolescence) built in, so LTS 2023 may not be patched anytime after May 2026.

| Considerations                       | Next + LTS | LTS Option A: LTS 2023 | LTS Option B: LTS Only |
|--------------------------------------|-----------:|-----------------------:|-----------------------:|
| Plugins backward bincompat           | ✅          | ✅                      | ✅                      |
| Scala 3 support post May 2026        |          ✅ |                      ❌ |                      ✅ |
| LTS 2025 library ecosystem           | ✅          | ❌                      | ✅                      |
| Backward sourcecompat with sbt 2.0.0 | ❌          | ✅                      | ❌                      |

So the situation you describe based on a somewhat hypothetical severe source-incompatibility cannot be avoided either with **Next + LTS** or **LTS Option B**.

> If sbt itself had an LTS that would use Scala LTS, that would be mitigated, but that is not the case, and I don't think there are enough resources on sbt to be able to do that.

Given that many plugins already are published against an arbitrary old version like sbt 1.5.8, I don't think it's a bad idea actually to designate a series as sbt LTS actually. I already maintain sbt 1.x series while developing sbt 2.x. If we dropped sbt 1.x eventually, the load won't be all that different from today.

**Comment by SethTisue** [comment]
I feel like there was a recent discussion where Eugene said sbt 2 would stay on Scala 3.7 and not move to 3.8 or 3.9, because 3.8 drops JDK 8 and JDK 11 support. Anyone remember where that discussion was?

**Reply by eed3si9n** [reply]
https://eed3si9n.com/sbt-2.0.0-RC8 I have a blurb in RC release note stating as such.

> sbt 2.0.0-RC7 uses Scala 3.7.4 in the metabuild. Rather than staying with Scala 3.3.x LTS (which will EOL in 2026), our current decision is to adopt the latest stable Scala 3.x versions built on JDK 8.

**Reply by SethTisue** [reply]
Eugene has now put a request-for-comment on this out to the community at https://users.scala-lang.org/t/rfc-sbt-2-0-on-jdk-17/12169

**Reply by eed3si9n** [reply]
Well I got talked into adopting JDK 17 by a group of mob on Discord, and I'm glad I did, since thus far the RFC comments have been overwhelmingly in favor of JDK 17 minimum.
